File: Chapter 1/Exercise1.1.jl
using PyPlot

function euler(f0::Float64, df::Float64, dt::Float64, tend::Float64)
    n::Int64 = Int(tend/dt) + 1
    f::Vector{Float64} = zeros(n)
    t::Vector{Float64} = zeros(n)
    for i = 1:n-1
        f[i+1] = f[i] + df*dt
        t[i+1] = t[i] + dt
    end
    return t,f
end

#=
function euler(f0, df, dt, tend)
    f = [f0]
    t = [0]
    n = Int(tend/dt)
    for i = 1:n
        push!(f, f[end] + df*dt)
        push!(t, t[end] + dt)
    end
    return t,f
end
=#

function main()
    t,v = euler(0.0, -9.8, 0.1, 10.0)
    clf()
    plot(t,v)
    xlabel("t")
    ylabel("v(t)")
    title("Chapter 1 Exercise 1.1")
    display(gcf())
end

# @time euler(0.0, -9.8, 1.0, 10.0)

# using the simple implementation of euler took ~70% more time
# than the slightly complicated implementation. The time and 
# memory allocation can be checked by commenting out the line 
# containing @time macro.

File: Chapter 1/Exercise1.1.py
import matplotlib.pyplot as plt

def df():
    return -9.8

def euler(f0, df, dt, tend):
    f = [f0]
    t = [0]
    n = int(tend/dt)
    for i in range(n):
        f += [f[-1] + df()*dt]
        t += [t[-1] + dt]
    return [t,f]

t,v = euler(0, df, 0.1, 10)
plt.plot(t,v)
plt.xlabel('$t$')
plt.ylabel('$v(t)$')
plt.axis([0, 10, -100, 0])
plt.tick_params(top=True, right=True, direction ='in')
plt.title('Chapter 1 Exercise 1.1')